---
title: "test"
author: "Dani Blumstein"
date: "2023-08-15"
output: html_document
---
```{r}
#BiocManager::install("impute")
library(tidyr)
library(dplyr)
library(ggbeeswarm)
library(lme4)
library(lmerTest)
#library(edgeR)
library(WGCNA)
library(cluster)
library(flashClust)
library(devtools)
library(Biobase)
library(preprocessCore)
library(gprofiler2)
library(doParallel)
library(readr)
```


```{r}
setwd("/Users/danielleblumstein/Documents/UNH/rnaseq/dehy_final")
countdata<-read.table(file = "data/gene.level.count.data.tsv", header = TRUE)
samples<- read_csv("results/samplespreadsheet.csv")
annos <- read_csv("/Users/danielleblumstein/Documents/UNH/rnaseq/dehy_final/data/annos.txt")
annos$ensid =  mapIds(org.Hs.eg.db,
                    keys=annos$gene, 
                    column="ENSEMBL",
                    keytype="SYMBOL",
                    multiVals="first")



kid_countdata <- dplyr::select(countdata, contains("kid"))
kid_samples <-samples %>% filter(tissue == "kid") 
kid_samples$trt <- as.factor(kid_samples$trt)
kid_samples$sex <- as.factor(kid_samples$sex)

kid_dds_trt <- DESeqDataSetFromMatrix(countData = kid_countdata,
                                       colData = kid_samples,
                                       design = ~ trt + sex)


#filter out genes where there are less than 8 samples with normalized counts greater than or equal to 10.
kid_dds_trt <- estimateSizeFactors(kid_dds_trt)
#idx <- rowSums(counts(kid_dds_trt, normalized=TRUE) >= 10 ) >= 8

#kid_dds_trt <- kid_dds_trt[idx,]
#print(paste("Number of genes after filtering by total expression:", nrow(kid_dds_trt)))
matrix <- as.data.frame(counts(kid_dds_trt))

#kid_dds_trt$trt_combo <- factor(paste0(kid_dds_trt$trt, kid_dds_trt$sex))
#design(kid_dds_trt) <- ~ pro_weight + trt_combo

kid_dds_trt <- DESeq(kid_dds_trt,test="Wald")
annos$ensembl <- as.character(annos$ensid)

#rename transcript rows to ensid names
#rownames(kid_dds_trt)=annos$ensid[match(names(kid_dds_trt),annos$gene)]
#remove the NAs (those transcripts dont have a ensid name)
#kid_dds_trt <- kid_dds_trt[complete.cases(rownames(kid_dds_trt)), ]
kid_dds_trt <- kid_dds_trt[!rownames(kid_dds_trt) == "NULL", ] 

kid_dds_trt <- estimateSizeFactors(kid_dds_trt)
normalized_counts <- vst(kid_dds_trt, blind = FALSE)
kid_Expr <- assay(normalized_counts)

kid <- heatmap(kid_Expr,Rowv=NA, col=rev(brewer.pal(9,"RdBu")))
```
## now you have normalized read counts - yay!


```{r}
##################  WGCNA  ##########################

#code to subset just DE genes
#kid_master <- read.csv("results/DE_kid/trt_genes.csv", header = TRUE)
#kidref <- kid_master$Gene
#kid_Expr <- subset(normalized_counts, subset = rownames(normalized_counts) %in% kidref)

#Remove gene information and transpose the expression data
rownames(kid_Expr)<-NULL
kid_Expr <- t(kid_Expr)

##########check included genes#########
gsg = goodSamplesGenes(kid_Expr, verbose = 3);
gsg$allOK
```

```{r}
# Trait data generation ---------------------------------------------------
# Now we read in the phenotypic trait data
kid_samples <- as.data.frame(kid_samples)
dim(kid_samples)
traitData_kid <- 'row.names<-'(kid_samples, kid_samples$sample)
traitData_kid$trt <- as.numeric(as.factor(traitData_kid$trt))
traitData_kid$sex <- as.numeric(as.factor(traitData_kid$sex))

# Choose columns of trait dataset to be used in downstream analyses
colnames(traitData_kid)
traitData_kid <- traitData_kid %>% dplyr::select(c('sex','total_weight_delta',"pro_weight","Na","BUN","AnGap","K","Crea","Hct","Cl","Glu","Hb*","TCO2","iCa","trt","mean_RQ","mean_EE","mean_H2Og","body_temp_3"))
traitData_kid <- as.data.frame(lapply(traitData_kid, as.numeric))
```

```{r}
#### filter samples ########
A_kid=adjacency(t(kid_Expr),type="distance")
# this calculates the whole network connectivity
k_kid=as.numeric(apply(A_kid,2,sum))-1
# standardized connectivity
Z.k_kid=scale(k_kid)

# Designate samples as outlying
# if their Z.k value is below the threshold
thresholdZ.k=-4 # often -2.5

# the color vector indicates outlyingness (red)
outlierColor_kid <- ifelse(Z.k_kid<thresholdZ.k,"red","black")

library(flashClust)
# calculate the cluster tree using flashClust or hclust
sampleTree_kid <- flashClust(as.dist(1-A_kid), method = "average")
# Convert traits to a color representation:
# where red indicates high values


traitColors = data.frame(numbers2colors(traitData_kid,signed=FALSE))
dimnames(traitColors)[[2]] = paste(names(traitData_kid))
datColors = data.frame(outlier = outlierColor_kid,traitColors)
plotDendroAndColors(sampleTree_kid,groupLabels=names(datColors),
                    colors=datColors,main="Sample Dendrogram and Trait Heatmap")
```


```{r}
############ Choosing the soft threshold beta via scale free topology
# Choose a set of soft thresholding powers
powers=c(1:30) # in practice this should include powers up to 20.
# choose power based on sft_JZ criterion
sft_kid=pickSoftThreshold(kid_Expr,powerVector=powers, networkType = "signed")
# Plot the results:

par(mfrow=c(1,2))
# sft_kid index as a function of different powers
plot(sft_kid$fitIndices[,1],-sign(sft_kid$fitIndices[,3])*sft_kid$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="sft_kid, unsigned R^2",type="n",main=paste("Scale independence"))
text(sft_kid$fitIndices[,1],-sign(sft_kid$fitIndices[,3])*sft_kid$fitIndices[,2],
     labels=powers,col="red", pch=1, cex=.50)
# this line corresponds to using an R^2 cut-off of h
abline(h=0.85,col="red")
# Mean connectivity as a function of different powers
plot(sft_kid$fitIndices[,1],sft_kid$fitIndices[,5],type="n",
     xlab="Soft Threshold (power)",ylab="Mean Connectivity",main=paste("Mean connectivity"))
text(sft_kid$fitIndices[,1],sft_kid$fitIndices[,5],labels=powers,col="red", cex=.50)

```

```{r}
##########I'm choosing 15 for my power based on scale independence and k for kid ###########

###########Automatic module detection via dynamic cutting
dim(kid_Expr)


mergingThresh = 0.25
net_kid = blockwiseModules(kid_Expr,corType="pearson",
                          maxBlockSize=14000,networkType="signed hybrid",power=21,minModuleSize=20,
                          mergeCutHeight=mergingThresh,numericLabels=F,saveTOMs=TRUE,
                          pamRespectsDendro=FALSE,saveTOMFileBase="kid_TEST")

```

```{r}
moduleLabelsAutomatic_kid=net_kid$colors

# Convert labels to colors for plotting
moduleColorsAutomatic_kid = labels2colors(moduleLabelsAutomatic_kid)

# A data frame with module eigengenes can be obtained as follows
MEsAutomatic_kid=net_kid$MEs

#this is proportional weight loss
weight_g_kid = as.data.frame(traitData_kid$pro_weight)
names(weight_g_kid)="weight_g_kid"

GS.kid=as.numeric(cor(kid_Expr,traitData_kid,use="p"))
# This translates the numeric values into colors
GS.kidWColor_kid=numbers2colors(GS.kid,signed=T)

blocknumber=1
datColors_kid=data.frame(moduleColorsAutomatic_kid,GS.kidWColor_kid)[net_kid$blockGenes[[blocknumber]],]

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net_kid$dendrograms[[blocknumber]],colors=datColors_kid,
                    groupLabels=c("Module colors"),dendroLabels=FALSE,
                    hang=0.03,addGuide=TRUE,guideHang=0.05)

# Choose a module assignment
moduleColors_kid=moduleColorsAutomatic_kid

# Define numbers of genes and samples
nGeneskid = ncol(kid_Expr)
nSampleskid = nrow(kid_Expr)

# Recalculate MEs with color labels
MEs0kid = moduleEigengenes(kid_Expr,moduleColors_kid)$eigengenes

MEs_kid = orderMEs(MEs0kid)
modTraitCor_kid = cor(MEs_kid, traitData_kid, use = "p")
modTraitP_kid = corPvalueStudent(modTraitCor_kid, nSampleskid)

#Since we have a moderately large number of modules and traits,
#a suitable graphical representation will help in reading
#the table. We color code each association by the correlation value:
# Will display correlations and their p-values
textMatrix_kid = paste(signif(modTraitCor_kid, 2), "\n(",
                      signif(modTraitP_kid, 1), ")", sep = "")
dim(textMatrix_kid) = dim(modTraitCor_kid)
par(mar = c(6, 8.5, 3, 3))

# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = modTraitCor_kid, xLabels = names(traitData_kid),
               yLabels = names(MEs_kid), ySymbols = names(MEs_kid), 
               colorLabels =FALSE,colors=blueWhiteRed(50),textMatrix=textMatrix_kid,
               setStdMargins = FALSE, cex.text = 0.4, zlim = c(-1,1),
               main = paste("Module-trait relationships, kid"))
png("figures/kid_WGCNA_heatmap.png", width = 9.5, height = 12, units = "in", res = 300)
```

mod merging
Now, using the new found measurements of dissimilarity, you can construct a cluster tree. You will also be adding a line at the height of .25. This height corresponds to a correlation of over 75%. Any branches below this line are more than 75% related, and you will thus be merging them!
```{r}
ME.dissimilarity = 1-cor(MEs0kid, use="complete") #Calculate eigengene dissimilarity
METree = hclust(as.dist(ME.dissimilarity), method = "average") #Clustering eigengenes 
par(mar = c(0,4,2,0)) #seting margin sizes
par(cex = 0.6);#scaling the graphic
plot(METree)
abline(h=.25, col = "red") #a height of .25 corresponds to correlation of .75

merge <- mergeCloseModules(kid_Expr, moduleColors_kid, cutHeight = .25)
# The merged module colors, assigning one color to each module
mergedColors = merge$colors
# Eigengenes of the new merged modules
mergedMEs = merge$newMEs

plotDendroAndColors(net_kid$dendrograms[[blocknumber]], cbind(moduleColors_kid, mergedColors), 
c("Original Module", "Merged Module"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main = "Gene dendrogram and module colors for original and merged modules")
```


```{r}
# Isolate weight from the clinical traits
pro_weight = as.data.frame(traitData_kid$pro_weight);
names(pro_weight) = "pro_weight"
mean_H2Og = as.data.frame(traitData_kid$mean_H2Og)
names(mean_H2Og) = "mean_H2Og"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs_kid, pro_weight,mean_H2Og))
# Plot the relationships among the eigengenes and the trait
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(5,4,1,2), cex.lab = 0.8, xLabelsAngle
= 90)
```



```{r}
############################write tables w/ corr info###################################################################
modNames_kid = substring(names(MEs_kid), 3)
geneModuleMembership_kid = as.data.frame(cor(kid_Expr, MEs_kid, use = "p"));
MMPvalue_kid = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership_kid), nSampleskid));
names(geneModuleMembership_kid) = paste("MM", modNames_kid, sep="");
names(MMPvalue_kid) = paste("p.MM", modNames_kid, sep="");
geneTraitSignificance_kid = as.data.frame(cor(kid_Expr, weight_g_kid, use = "p"));
GSPvalue_kid = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance_kid), nSampleskid));
names(geneTraitSignificance_kid) = paste("GS.", names(weight_g_kid), sep="");
names(GSPvalue_kid) = paste("p.GS.", names(weight_g_kid), sep="");

All_module_correlations_kid <- subset(signif(modTraitP_kid, 1))
W_module_correlations_kid <- All_module_correlations_kid[,1]
W_module_adjusted_correlations_kid <- p.adjust(W_module_correlations_kid, method = "fdr")

# (aka. module eigengene based connectivity kME):
datKME_kid=signedKME(kid_Expr, MEs_kid)
melt(datKME_kid)


Module_eigengenes_kid <- net_kid$MEs
rownames(MEs_kid) <- rownames(kid_Expr)
write.csv(MEs_kid, file="results/WGCNAresults/ALLSAMPLES_kid_MEs_signed.csv")
write.csv(modTraitP_kid, file="results/WGCNAresults/ALLSAMPLES_kid_MEs_P_Pearson.csv")
```

```{r}
for (row in 1:nrow(modTraitP_kid)){
    rownames(modTraitP_kid)[row] <-  sub("ME", "", rownames(modTraitP_kid)[row])
}
 
color_counts <- as.data.frame(table(moduleColorsAutomatic_kid))
modTraitP_kid_df <- as.data.frame(modTraitP_kid)
modTraitP_kid_df <- tibble::rownames_to_column(modTraitP_kid_df, "moduleColorsAutomatic_kid")

merge(modTraitP_kid_df, color_counts, by="moduleColorsAutomatic_kid", modTraitP_kid_df=TRUE)
  
nodes <- as.data.frame(table(moduleColorsAutomatic_kid))
links <- melt(ME.dissimilarity)
data_zero <- filter_if(links, is.numeric, all_vars((.) != 0))

dim(ME.dissimilarity)
net <- graph_from_adjacency_matrix(ME.dissimilarity,weighted = TRUE,)
net <- simplify(net, remove.multiple = F, remove.loops = T)
V(net)$color <- rownames(ME.dissimilarity)
V(net)$size <- log(nodes$Freq)*5
E(net)$width <- log(data_zero$value)*10
#change arrow size and edge color:
E(net)$edge.color <- "gray80"
E(net)$edge.arrow.mode <- 0

layouts <- grep("^layout_", ls("package:igraph"), value=TRUE)[-1]
# Remove layouts that do not apply to our graph.
layouts <- layouts[!grepl("bipartite|merge|norm|sugiyama|tree", layouts)]
par(mfrow=c(3,3), mar=c(1,1,1,1))
for (layout in layouts) {
print(layout)
l <- do.call(layout, list(net))
plot(net, edge.arrow.mode=0, layout=l, main=layout) }
# Choose final layout
par(mfrow=c(1,1))
plot(net, edge.arrow.mode=0, layout=layout_with_kk)

#barplot(nodes$Freq, col = rownames(ME.dissimilarity), ylim = c(0,5000))
```


```{r}
############ kid models ##############
library(lme4)
library(lmerTest)

traitData_kid <- kid_samples
traitData_kid <- 'row.names<-'(kid_samples, kid_samples$sample)
traitData_kid$trt <- as.numeric(as.factor(traitData_kid$trt))
traitData_kid$sex <- as.numeric(as.factor(traitData_kid$sex))

# Choose columns of trait dataset to be used in downstream analyses
colnames(traitData_kid)
traitData_kid <- traitData_kid %>% select(c('sex','total_weight_delta',"pro_weight","Na","BUN","AnGap","K","Crea","Hct","Cl","Glu","Hb*","TCO2","iCa","trt","mean_RQ","mean_EE","mean_H2Og","body_temp_3","sample"))

MEs_kid<- read.csv("results/WGCNAresults/ALLSAMPLES_kid_MEs_signed.csv", header = TRUE)
MEs_kid$X <-  sub("X*", "", MEs_kid$X)

pheno_ME_kid <- merge(MEs_kid, traitData_kid, by.x = "X", by.y = "sample")

plot(pheno_ME_kid$MEgreen ~ pheno_ME_kid$trt, col = pheno_ME_kid$trt)
```


```{r}
##################################write MM and GS values to file for each gene
# Create the starting data frame

genes_kid <- assay(normalized_counts)
geneInfo0_kid = data.frame(Gene = genes_kid,
                          moduleColor = moduleColors_kid,
                          geneTraitSignificance_kid,
                          GSPvalue_kid)
# Order modules by their significance for weight
MEs_kid2 <- MEs_kid[,-1]
modOrder = order(-abs(cor(MEs_kid2, weight_g_kid, use = "p")));
# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership_kid))
{
  oldNames = names(geneInfo0_kid)
  geneInfo0_kid = data.frame(geneInfo0_kid, geneModuleMembership_kid[, modOrder[mod]],
                            MMPvalue_kid[, modOrder[mod]]);
  names(geneInfo0_kid) = c(oldNames, paste("MM.", modNames_kid[modOrder[mod]], sep=""),
                          paste("p.MM.", modNames_kid[modOrder[mod]], sep=""))
}

#geneInfo0_kid$ensembl <- rownames(geneInfo0_kid)
#geneInfo0_kid <- merge(geneInfo0_kid, annos, by.x = "ensembl", by.y = "ensembl")
geneInfo0_kid$ensid <- NULL

write.csv(geneInfo0_kid, file="results/WGCNAresults/ALLSAMPLES_kid_wgcna_gene_module_membership_pearson_signed.csv")
```


GO
```{r}
mods <- unique(geneInfo0_kid$moduleColor)
#removing colors getting stuck on FDR = 0.05
#mods <- mods[- 4] 
#mods <- mods[- 5]  
#mods <- mods[- 9] 
#mods <- mods[- 10]
#mods <- mods[- 10]
#mods <- mods[- 10]
#mods <- mods[- 11]

for (color in mods) 
  {
wgcna <- geneInfo0_kid$ensembl[which(geneInfo0_kid$moduleColor==color)]
kid_dds_trt1 <- subset(kid_dds_trt, rownames(kid_dds_trt) %in% wgcna)


restrt <- results(kid_dds_trt1, alpha = 0.01, contrast = c("trt", "yes", "no"))
restrt$transcript <- mcols(restrt)$transcript
restrt$symbol <- mcols(restrt)$symbol

restrt$ensembl <- sapply( strsplit( rownames(restrt), split="\\+" ), "[", 1 )
#ensembl = useMart( "ensembl", dataset = "hsapiens_gene_ensembl" )

#querying with the ENSEMBL gene id and requesting the Entrez gene id and HGNC gene symbol.
genemap <- getBM( attributes = c("ensembl_gene_id", "hgnc_symbol"),
                  filters = "ensembl_gene_id",
                  values = rownames(restrt),
                  mart = ensembl)

idx <- match(rownames(restrt), genemap$ensembl_gene_id )
restrt$hgnc_symbol <- genemap$hgnc_symbol[ idx ]
hgnc_symbol <- restrt$hgnc_symbol

mcols(kid_dds_trt1) <- cbind(mcols(kid_dds_trt1), hgnc_symbol)
restrt$hgnc_symbol <- rowData(kid_dds_trt1)$hgnc_symbol

kid_dds_trt1 <- estimateSizeFactors(kid_dds_trt1)

anno_df <- data.frame(gene_id = restrt$ensembl, gene_name = restrt$hgnc_symbol)
vst_macrophage <- varianceStabilizingTransformation(kid_dds_trt1)

bg_ids <- rownames(kid_dds_trt1)[rowSums(counts(kid_dds_trt1)) > 0]
bg_symbols <- mapIds(org.Hs.eg.db,
                     keys = bg_ids,
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")
 
# res_enrich object
#library("AnnotationDbi")
de_symbols_IFNg_vs_naive <- deseqresult2df(restrt, FDR =0.05)$hgnc_symbol

# res_enrich object ------------------------------------------------------------
 topgoDE <-
  pcaExplorer::topGOtable(de_symbols_IFNg_vs_naive,
                          bg_symbols,
                          ontology = "BP",
                          mapping = "org.Hs.eg.db",
                          geneID = "symbol",
                          topTablerows = 30)

res_enrich <- shake_topGOtableResult(topgoDE)
res_enrich_mod <- get_aggrscores(res_enrich = res_enrich,
                                        res_de = restrt,
                                        annotation_obj = anno_df,
                                        aggrfun = mean)
res_enrich_mod$mod <- color
write.csv(res_enrich_mod, file=paste0("results/WGCNAresults/kid/kid_wgcna_GO",color,".csv"))
}


```


