---
title: "WGCNA_lu"
author: "Dani Blumstein"
date: "2023-08-01"
output: html_document
---

```{r}
#BiocManager::install("impute")
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggbeeswarm)
library(lme4)
library(lmerTest)
#library(edgeR)
library(WGCNA)
library(cluster)
library(flashClust)
library(devtools)
library(Biobase)
library(preprocessCore)
library(gprofiler2)
library(doParallel)
```


```{r}
setwd("/Users/danielleblumstein/Documents/UNH/rnaseq/dehy_final")
countdata<-read.table(file = "data/gene.level.count.data.tsv", header = TRUE)
samples<- read_csv("results/samplespreadsheet.csv")
annos <- read_csv("/Users/danielleblumstein/Documents/UNH/rnaseq/dehy_final/data/annos.txt")
annos$ensid = mapIds(org.Hs.eg.db,
                    keys=annos$gene, 
                    column="ENSEMBL",
                    keytype="SYMBOL",
                    multiVals="first")



lu_countdata <- dplyr::select(countdata, contains("lu"))
lu_samples <-samples %>% filter(tissue == "lu") 
lu_samples$trt <- as.factor(lu_samples$trt)
lu_samples$sex <- as.factor(lu_samples$sex)

lu_dds_trt <- DESeqDataSetFromMatrix(countData = lu_countdata,
                                       colData = lu_samples,
                                       design = ~ trt + sex)


#filter out genes where there are less than 8 samples with normalized counts greater than or equal to 10.
lu_dds_trt <- estimateSizeFactors(lu_dds_trt)
idx <- rowSums(counts(lu_dds_trt, normalized=TRUE) >= 10 ) >= 8

lu_dds_trt <- lu_dds_trt[idx,]
print(paste("Number of genes after filtering by total expression:", nrow(lu_dds_trt)))
matrix <- as.data.frame(counts(lu_dds_trt))

#lu_dds_trt$trt_combo <- factor(paste0(lu_dds_trt$trt, lu_dds_trt$sex))
#design(lu_dds_trt) <- ~ pro_weight + trt_combo

lu_dds_trt <- DESeq(lu_dds_trt,test="Wald")
annos$ensembl <- as.character(annos$ensid)

#rename transcript rows to ensid names
#rownames(lu_dds_trt)=annos$ensid[match(names(lu_dds_trt),annos$gene)]
#remove the NAs (those transcripts dont have a ensid name)
#lu_dds_trt <- lu_dds_trt[complete.cases(rownames(lu_dds_trt)), ]
#lu_dds_trt <- lu_dds_trt[!rownames(lu_dds_trt) == "NULL", ] 

lu_dds_trt <- estimateSizeFactors(lu_dds_trt)
normalized_counts <- vst(lu_dds_trt, blind = FALSE)
lu_Expr <- assay(normalized_counts)
```
## now you have normalized read counts - yay!

```{r}
##################  WGCNA  ##########################

#code to subset just DE genes
#lu_master <- read.csv("results/DE_lu/trt_genes.csv", header = TRUE)
#luref <- lu_master$Gene
#lu_Expr <- subset(normalized_counts, subset = rownames(normalized_counts) %in% luref)

#Remove gene information and transpose the expression data
rownames(lu_Expr)<-NULL
lu_Expr <- t(lu_Expr)

##########check included genes#########
gsg = goodSamplesGenes(lu_Expr, verbose = 3);
gsg$allOK
```

```{r}
# Trait data generation ---------------------------------------------------
# Now we read in the phenotypic trait data
lu_samples <- as.data.frame(lu_samples)
dim(lu_samples)
traitData_lu <- 'row.names<-'(lu_samples, lu_samples$sample)
traitData_lu$trt <- as.numeric(as.factor(traitData_lu$trt))
traitData_lu$sex <- as.numeric(as.factor(traitData_lu$sex))

# Choose columns of trait dataset to be used in downstream analyses
colnames(traitData_lu)
traitData_lu <- traitData_lu %>% select(c('sex','total_weight_delta',"pro_weight","Na","BUN","AnGap","K","Crea","Hct","Cl","Glu","Hb*","TCO2","iCa","trt","mean_RQ","mean_EE","mean_H2Og","body_temp_3"))
#traitData_lu <- 'row.names<-'(traitData_lu, traitData_lu$sample)
traitData_lu <- as.data.frame(lapply(traitData_lu, as.numeric))
```

```{r}
#### filter samples ########
A_lu=adjacency(t(lu_Expr),type="distance")
# this calculates the whole network connectivity
k_lu=as.numeric(apply(A_lu,2,sum))-1
# standardized connectivity
Z.k_lu=scale(k_lu)

# Designate samples as outlying
# if their Z.k value is below the threshold
thresholdZ.k=-4 # often -2.5

# the color vector indicates outlyingness (red)
outlierColor_lu <- ifelse(Z.k_lu<thresholdZ.k,"red","black")

# calculate the cluster tree using flashClust or hclust
sampleTree_lu <- flashClust(as.dist(1-A_lu), method = "average")
# Convert traits to a color representation:
# where red indicates high values


traitColors = data.frame(numbers2colors(traitData_lu,signed=FALSE))
dimnames(traitColors)[[2]] = paste(names(traitData_lu))
datColors = data.frame(outlier = outlierColor_lu,traitColors)
plotDendroAndColors(sampleTree_lu,groupLabels=names(datColors),
                    colors=datColors,main="Sample Dendrogram and Trait Heatmap")

```


```{r}
############ Choosing the soft threshold beta via scale free topology
# Choose a set of soft thresholding powers
powers=c(1:30) # in practice this should include powers up to 20.
# choose power based on sft_JZ criterion
sft_lu=pickSoftThreshold(lu_Expr,powerVector=powers, networkType = "signed")
# Plot the results:

par(mfrow=c(1,2))
# sft_lu index as a function of different powers
plot(sft_lu$fitIndices[,1],-sign(sft_lu$fitIndices[,3])*sft_lu$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="sft_lu, unsigned R^2",type="n",main=paste("Scale independence"))
text(sft_lu$fitIndices[,1],-sign(sft_lu$fitIndices[,3])*sft_lu$fitIndices[,2],
     labels=powers,col="red", pch=1, cex=.50)
# this line corresponds to using an R^2 cut-off of h
abline(h=0.85,col="red")
# Mean connectivity as a function of different powers
plot(sft_lu$fitIndices[,1],sft_lu$fitIndices[,5],type="n",
     xlab="Soft Threshold (power)",ylab="Mean Connectivity",main=paste("Mean connectivity"))
text(sft_lu$fitIndices[,1],sft_lu$fitIndices[,5],labels=powers,col="red", cex=.50)
```

```{r}
##########I'm choosing 15 for my power based on scale independence and k for lu ###########

###########Automatic module detection via dynamic cutting

merlungThresh = 0.25
net_lu = blockwiseModules(lu_Expr,corType="pearson",
                          maxBlockSize=14000,networkType="signed hybrid",power=20,minModuleSize=20,
                          mergeCutHeight=merlungThresh,numericLabels=TRUE,saveTOMs=TRUE,
                          pamRespectsDendro=FALSE,saveTOMFileBase="lu_TEST")

moduleLabelsAutomatic_lu=net_lu$colors

# Convert labels to colors for plotting
moduleColorsAutomatic_lu = labels2colors(moduleLabelsAutomatic_lu)

# A data frame with module eigengenes can be obtained as follows
MEsAutomatic_lu=net_lu$MEs
```

```{r}
#this is proportional weight loss
weight_g_lu = as.data.frame(traitData_lu$pro_weight)
names(weight_g_lu)="weight_g_lu"

GS.lu=as.numeric(cor(lu_Expr,traitData_lu,use="p"))
# This translates the numeric values into colors
GS.luWColor_lu=numbers2colors(GS.lu,signed=T)

blocknumber=1
datColors_lu=data.frame(moduleColorsAutomatic_lu,GS.luWColor_lu)[net_lu$blockGenes[[blocknumber]],]

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net_lu$dendrograms[[blocknumber]],colors=datColors_lu,
                    groupLabels=c("Module colors"),dendroLabels=FALSE,
                    hang=0.03,addGuide=TRUE,guideHang=0.05)

# Choose a module assignment
moduleColors_lu=moduleColorsAutomatic_lu

# Define numbers of genes and samples
nGeneslu = ncol(lu_Expr)
nSampleslu = nrow(lu_Expr)

# Recalculate MEs with color labels
MEs0lu = moduleEigengenes(lu_Expr,moduleColors_lu)$eigengenes

MEs_lu = orderMEs(MEs0lu)
modTraitCor_lu = cor(MEs_lu, traitData_lu, use = "p")
modTraitP_lu = corPvalueStudent(modTraitCor_lu, nSampleslu)

#Since we have a moderately large number of modules and traits,
#a suitable graphical representation will help in reading
#the table. We color code each association by the correlation value:
# Will display correlations and their p-values
textMatrix_lu = paste(signif(modTraitCor_lu, 2), "\n(",
                      signif(modTraitP_lu, 1), ")", sep = "")
dim(textMatrix_lu) = dim(modTraitCor_lu)
par(mar = c(6, 8.5, 3, 3))

# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = modTraitCor_lu, xLabels = names(traitData_lu),
               yLabels = names(MEs_lu), ySymbols = names(MEs_lu), 
               colorLabels =FALSE,colors=blueWhiteRed(50),textMatrix=textMatrix_lu,
               setStdMarluns = FALSE, cex.text = 0.4, zlim = c(-1,1),
               main = paste("Module-trait relationships, lu"))
png("figures/lu_WGCNA_heatmap.png", width = 9.5, height = 12, units = "in", res = 300)
```

```{r}
############################write tables w/ corr info###################################################################
modNames_lu = substring(names(MEs_lu), 3)
geneModuleMembership_lu = as.data.frame(cor(lu_Expr, MEs_lu, use = "p"));
MMPvalue_lu = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership_lu), nSampleslu));
names(geneModuleMembership_lu) = paste("MM", modNames_lu, sep="");
names(MMPvalue_lu) = paste("p.MM", modNames_lu, sep="");
geneTraitSignificance_lu = as.data.frame(cor(lu_Expr, weight_g_lu, use = "p"));
GSPvalue_lu = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance_lu), nSampleslu));
names(geneTraitSignificance_lu) = paste("GS.", names(weight_g_lu), sep="");
names(GSPvalue_lu) = paste("p.GS.", names(weight_g_lu), sep="");

All_module_correlations_lu <- subset(signif(modTraitP_lu, 1))
W_module_correlations_lu <- All_module_correlations_lu[,1]
W_module_adjusted_correlations_lu <- p.adjust(W_module_correlations_lu, method = "fdr")

# (aka. module eigengene based connectivity kME):
datKME_lu=signedKME(lu_Expr, MEs_lu)

Module_eigengenes_lu <- net_lu$MEs
rownames(MEs_lu) <- rownames(lu_Expr)
write.csv(MEs_lu, file="results/WGCNAresults/ALLSAMPLES_lu_MEs_signed.csv")
write.csv(modTraitP_lu, file="results/WGCNAresults/ALLSAMPLES_lu_MEs_P_Pearson.csv")
```

```{r}
############ lu models ##############
library(lme4)
library(lmerTest)

traitData_lu <- lu_samples
traitData_lu <- 'row.names<-'(lu_samples, lu_samples$sample)
traitData_lu$trt <- as.numeric(as.factor(traitData_lu$trt))
traitData_lu$sex <- as.numeric(as.factor(traitData_lu$sex))

# Choose columns of trait dataset to be used in downstream analyses
colnames(traitData_lu)
traitData_lu <- traitData_lu %>% select(c('sex','total_weight_delta',"pro_weight","Na","BUN","AnGap","K","Crea","Hct","Cl","Glu","Hb*","TCO2","iCa","trt","mean_RQ","mean_EE","mean_H2Og","body_temp_3","sample"))

MEs_lu<- read.csv("results/WGCNAresults/ALLSAMPLES_lu_MEs_signed.csv", header = TRUE)
MEs_lu$X <-  sub("X*", "", MEs_lu$X)

pheno_ME_lu <- merge(MEs_lu, traitData_lu, by.x = "X", by.y = "sample")

plot(pheno_ME_lu$MEgreen ~ pheno_ME_lu$trt, col = pheno_ME_lu$trt)
```


```{r}
##################################write MM and GS values to file for each gene
# Create the starting data frame

genes_lu <- assay(normalized_counts)
geneInfo0_lu = data.frame(Gene = genes_lu,
                          moduleColor = moduleColors_lu,
                          geneTraitSignificance_lu,
                          GSPvalue_lu)
# Order modules by their significance for weight
MEs_lu2 <- MEs_lu[,-1]
modOrder = order(-abs(cor(MEs_lu2, weight_g_lu, use = "p")));
# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership_lu))
{
  oldNames = names(geneInfo0_lu)
  geneInfo0_lu = data.frame(geneInfo0_lu, geneModuleMembership_lu[, modOrder[mod]],
                            MMPvalue_lu[, modOrder[mod]]);
  names(geneInfo0_lu) = c(oldNames, paste("MM.", modNames_lu[modOrder[mod]], sep=""),
                          paste("p.MM.", modNames_lu[modOrder[mod]], sep=""))
}

geneInfo0_lu$ensembl <- rownames(geneInfo0_lu)
geneInfo0_lu <- merge(geneInfo0_lu, annos, by.x = "ensembl", by.y = "ensembl")
geneInfo0_lu$ensid <- NULL

write.csv(geneInfo0_lu, file="results/WGCNAresults/ALLSAMPLES_lu_wgcna_gene_module_membership_pearson_signed.csv")
```


GO
```{r}
mods <- unique(geneInfo0_lu$moduleColor)
#removing colors getting stuck on FDR = 0.05
mods <- mods[- 11]

for (color in mods) 
  {
wgcna <- geneInfo0_lu$ensembl[which(geneInfo0_lu$moduleColor==color)]
lu_dds_trt1 <- subset(lu_dds_trt, rownames(lu_dds_trt) %in% wgcna)


restrt <- results(lu_dds_trt1, alpha = 0.01, contrast = c("trt", "yes", "no"))
restrt$transcript <- mcols(restrt)$transcript
restrt$symbol <- mcols(restrt)$symbol

restrt$ensembl <- sapply( strsplit( rownames(restrt), split="\\+" ), "[", 1 )
#ensembl = useMart( "ensembl", dataset = "hsapiens_gene_ensembl" )

#querying with the ENSEMBL gene id and requesting the Entrez gene id and HGNC gene symbol.
genemap <- getBM( attributes = c("ensembl_gene_id", "hgnc_symbol"),
                  filters = "ensembl_gene_id",
                  values = rownames(restrt),
                  mart = ensembl)

idx <- match(rownames(restrt), genemap$ensembl_gene_id )
restrt$hgnc_symbol <- genemap$hgnc_symbol[ idx ]
hgnc_symbol <- restrt$hgnc_symbol

mcols(lu_dds_trt1) <- cbind(mcols(lu_dds_trt1), hgnc_symbol)
restrt$hgnc_symbol <- rowData(lu_dds_trt1)$hgnc_symbol

lu_dds_trt1 <- estimateSizeFactors(lu_dds_trt1)

anno_df <- data.frame(gene_id = restrt$ensembl, gene_name = restrt$hgnc_symbol)
vst_macrophage <- varianceStabilizingTransformation(lu_dds_trt1)

bg_ids <- rownames(lu_dds_trt1)[rowSums(counts(lu_dds_trt1)) > 0]
bg_symbols <- mapIds(org.Hs.eg.db,
                     keys = bg_ids,
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")
 
# res_enrich object
#library("AnnotationDbi")
de_symbols_IFNg_vs_naive <- deseqresult2df(restrt, FDR =0.05)$hgnc_symbol

# res_enrich object ------------------------------------------------------------
 topgoDE <-
  pcaExplorer::topGOtable(de_symbols_IFNg_vs_naive,
                          bg_symbols,
                          ontology = "BP",
                          mapping = "org.Hs.eg.db",
                          geneID = "symbol",
                          topTablerows = 30)

res_enrich <- shake_topGOtableResult(topgoDE)
res_enrich_mod <- get_aggrscores(res_enrich = res_enrich,
                                        res_de = restrt,
                                        annotation_obj = anno_df,
                                        aggrfun = mean)
res_enrich_mod$mod <- color
write.csv(res_enrich_mod, file=paste0("results/WGCNAresults/lu/lu_wgcna_GO",color,".csv"))
}
```